TiDB panics when calling RegionStore.accessStore

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
I'm not able to reproduce this issue.

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
TiDB runs normally

### 3. What did you see instead (Required)
TiDB panics when calling RegionStore.accessStore
```
[2020/09/23 09:15:43.977 +00:00] [INFO] [region_cache.go:829] ["switch region leader to specific leader due to kv return NotLeader"] [regionID=47144] [currIdx=0] [leaderStoreID=66]
[2020/09/23 09:15:43.977 +00:00] [INFO] [region_cache.go:609] ["switch region peer to next due to send request fail"] [current="region ID: 50568, meta: id:50568 start_key:\"t\\200\\000\\000\\000\\000\\000\\0001_i\\200\\000\\000\\000\\000\\000\\000\\001\\00186864293\\377-f4f6-59\\377f6-9d07-\\37731903362\\3778caf\\000\\000\\000\\000\\373\" end_key:\"t\\200\\000\\000\\000\\000\\000\\0001_i\\200\\000\\000\\000\\000\\000\\000\\001\\001869cf186\\377-162e-5c\\377bd-a0f0-\\3777549f6c7\\3779f13\\000\\000\\000\\000\\373\" region_epoch:<conf_ver:5 version:36 > peers:<id:50570 store_id:4 > peers:<id:50571 store_id:66 > , peer: id:50571 store_id:66 , addr: db-tikv-2.db-tikv-peer.tidb1306611994220040192.svc:20160, idx: 1, reqStoreType: TiKvOnly, runStoreType: tikv"] [needReload=false] [error="wait recvLoop: context deadline exceeded"]
[2020/09/23 09:15:43.978 +00:00] [INFO] [region_cache.go:414] ["invalidate current region, because others failed on same store"] [region=5600] [store=db-tikv-2.db-tikv-peer.tidb1306611994220040192.svc:20160]
[2020/09/23 09:15:43.980 +00:00] [INFO] [region_cache.go:829] ["switch region leader to specific leader due to kv return NotLeader"] [regionID=11412] [currIdx=0] [leaderStoreID=66]
[2020/09/23 09:15:43.981 +00:00] [INFO] [region_cache.go:1630] ["[liveness] request kv status fail"] [store=0.0.0.0:20180] [error="Get https://0.0.0.0:20180/status: dial tcp 0.0.0.0:20180: connect: connection refused"]
panic: runtime error: index out of range [2] with length 2

goroutine 156282637 [running]:
github.com/pingcap/tidb/store/tikv.(*RegionStore).accessStore(...)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/region_cache.go:117
github.com/pingcap/tidb/store/tikv.(*RegionCache).OnSendFail(0xc0001ba2a0, 0xc18dc16b80, 0xc13ca4ed70, 0xc17619a000, 0x36cf480, 0xc1f42b9e40)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/region_cache.go:578 +0x93a
github.com/pingcap/tidb/store/tikv.(*RegionRequestSender).onSendFail(0xc0b05ae050, 0xc18dc16b80, 0xc13ca4ed70, 0x36cf480, 0xc1f42b9e40, 0xc13c675220, 0x4a817c800)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/region_request.go:455 +0x196
github.com/pingcap/tidb/store/tikv.(*RegionRequestSender).sendReqToRegion(0xc0b05ae050, 0xc18dc16b80, 0xc13ca4ed70, 0xc13c675220, 0x4a817c800, 0x0, 0x0, 0x36cf480, 0xc1f42b9e40)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/region_request.go:404 +0x261
github.com/pingcap/tidb/store/tikv.(*RegionRequestSender).SendReqCtx(0xc0b05ae050, 0xc18dc16b80, 0xc13c675220, 0x1878, 0x5, 0x23, 0x4a817c800, 0x15c3800, 0xc13c904f70, 0xc1a67ca410, ...)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/region_request.go:293 +0x2d0
github.com/pingcap/tidb/store/tikv.(*RegionRequestSender).SendReq(...)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/region_request.go:202
github.com/pingcap/tidb/store/tikv.actionCommit.handleSingleBatch(0x1, 0xc1afd88420, 0xc18dc16b80, 0x1878, 0x5, 0x23, 0xc1b1184254, 0x1, 0x353, 0xc1b1262df8, ...)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:1076 +0x285
github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnBatches(0xc1afd88420, 0xc18dc16b80, 0x37180a0, 0xc13c761e88, 0xc1bce2d580, 0x1, 0x1, 0xc1b1184254, 0x1)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:628 +0x110
github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnGroupMutations(0xc1afd88420, 0xc18dc16b80, 0x37180a0, 0xc13c761e88, 0xc1bce2d500, 0x1, 0x1, 0x353, 0xc1b128adf8)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:609 +0x661
github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnMutations(0xc1afd88420, 0xc18dc16b80, 0x37180a0, 0xc13c761e88, 0xc1b1184254, 0x1, 0x353, 0xc1b1262df8, 0x1, 0x353, ...)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:478 +0x1c2
github.com/pingcap/tidb/store/tikv.actionCommit.handleSingleBatch(0x1, 0xc1afd88420, 0xc18dc16b80, 0x1878, 0x5, 0x23, 0xc1b1184254, 0x1, 0x353, 0xc1b1262df8, ...)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:1100 +0x19aa
github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnBatches(0xc1afd88420, 0xc18dc16b80, 0x37180a0, 0xc16d12d2a8, 0xc1346efa00, 0x1, 0x1, 0xc1b1184254, 0x1)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:628 +0x110
github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnGroupMutations(0xc1afd88420, 0xc18dc16b80, 0x37180a0, 0xc16d12d2a8, 0xc1346ef980, 0x1, 0x1, 0x353, 0xc1b128adf8)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:609 +0x661
github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnMutations(0xc1afd88420, 0xc18dc16b80, 0x37180a0, 0xc16d12d2a8, 0xc1b1184254, 0x1, 0x353, 0xc1b1262df8, 0x1, 0x353, ...)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:478 +0x1c2
github.com/pingcap/tidb/store/tikv.actionCommit.handleSingleBatch(0x1, 0xc1afd88420, 0xc18dc16b80, 0x1878, 0x5, 0x23, 0xc1b1184254, 0x1, 0x353, 0xc1b1262df8, ...)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:1100 +0x19aa
github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnBatches(0xc1afd88420, 0xc18dc16b80, 0x37180a0, 0xc0de7f6648, 0xc0941a5700, 0x1, 0x1, 0xc1b1184254, 0x1)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:628 +0x110
github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnGroupMutations(0xc1afd88420, 0xc18dc16b80, 0x37180a0, 0xc0de7f6648, 0xc0941a5680, 0x1, 0x1, 0x353, 0xc1b128adf8)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:609 +0x661
github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnMutations(0xc1afd88420, 0xc18dc16b80, 0x37180a0, 0xc0de7f6648, 0xc1b1184254, 0x1, 0x353, 0xc1b1262df8, 0x1, 0x353, ...)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:478 +0x1c2
github.com/pingcap/tidb/store/tikv.actionCommit.handleSingleBatch(0x309b200, 0xc1afd88420, 0xc18dc16b80, 0x1878, 0x5, 0x23, 0xc1b1184254, 0x1, 0x353, 0xc1b1262df8, ...)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:1100 +0x19aa
github.com/pingcap/tidb/store/tikv.(*batchExecutor).startWorker.func1(0xc2631814c0, 0xc016cb3080, 0x1878, 0x5, 0x23, 0xc1b1184254, 0x1, 0x353, 0xc1b1262df8, 0x1, ...)
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:1547 +0x17c
created by github.com/pingcap/tidb/store/tikv.(*batchExecutor).startWorker
	/home/jenkins/agent/workspace/tidb_v4.0.6/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:1529 +0x1b7

```

### 4. What is your TiDB version? (Required)
`v4.0.6`
<!-- Paste the output of SELECT tidb_version() -->

--- a/store/tikv/region_cache.go
+++ b/store/tikv/region_cache.go
@@ -1221,7 +1221,7 @@ func (r *Region) GetMeta() *metapb.Region {
 // GetLeaderPeerID returns leader peer ID.
 func (r *Region) GetLeaderPeerID() uint64 {
 	store := r.getStore()
-	if int(store.workTiKVIdx) >= len(r.meta.Peers) {
+	if int(store.workTiKVIdx) >= store.accessStoreNum(TiKvOnly) {
 		return 0
 	}
 	storeIdx, _ := store.accessStore(TiKvOnly, store.workTiKVIdx)
@@ -1231,7 +1231,7 @@ func (r *Region) GetLeaderPeerID() uint64 {
 // GetLeaderStoreID returns the store ID of the leader region.
 func (r *Region) GetLeaderStoreID() uint64 {
 	store := r.getStore()
-	if int(store.workTiKVIdx) >= len(r.meta.Peers) {
+	if int(store.workTiKVIdx) >= store.accessStoreNum(TiKvOnly) {
 		return 0
 	}
 	storeIdx, _ := store.accessStore(TiKvOnly, store.workTiKVIdx)

--- a/store/tikv/region_cache.go
+++ b/store/tikv/region_cache.go
@@ -554,6 +554,16 @@ func (c *RegionCache) OnSendFail(bo *Backoffer, ctx *RPCContext, scheduleReload 
 	tikvRegionCacheCounterWithSendFail.Inc()
 	r := c.getCachedRegionWithRLock(ctx.Region)
 	if r != nil {
+		peersNum := len(r.meta.Peers)
+		if len(ctx.Meta.Peers) != peersNum {
+			logutil.Logger(bo.ctx).Info("retry and refresh current ctx after send request fail and up/down stores length changed",
+				zap.Stringer("current", ctx),
+				zap.Bool("needReload", scheduleReload),
+				zap.Reflect("oldPeers", ctx.Meta.Peers),
+				zap.Reflect("newPeers", r.meta.Peers),
+				zap.Error(err))
+			return
+		}
 		rs := r.getStore()
 		if err != nil {
 			storeIdx, s := rs.accessStore(ctx.AccessMode, ctx.AccessIdx)

--- a/store/tikv/region_cache_test.go
+++ b/store/tikv/region_cache_test.go
@@ -28,6 +28,7 @@ import (
 	"github.com/pingcap/kvproto/pkg/metapb"
 	"github.com/pingcap/tidb/kv"
 	"github.com/pingcap/tidb/store/mockstore/mocktikv"
+	pd "github.com/tikv/pd/client"
 )
 
 type testRegionCacheSuite struct {
@@ -1077,6 +1078,42 @@ func (s *testRegionCacheSuite) TestMixedMeetEpochNotMatch(c *C) {
 	regionErr := &errorpb.Error{EpochNotMatch: &errorpb.EpochNotMatch{}}
 	reqSend.onRegionError(s.bo, ctxFollower1, &followReqSeed, regionErr)
 	c.Assert(followReqSeed, Equals, uint32(1))
+}
+
+func (s *testRegionCacheSuite) TestPeersLenChange(c *C) {
+	// 2 peers [peer1, peer2] and let peer2 become leader
+	loc, err := s.cache.LocateKey(s.bo, []byte("a"))
+	c.Assert(err, IsNil)
+	s.cache.UpdateLeader(loc.Region, s.store2, 0)
+
+	// current leader is peer2 in [peer1, peer2]
+	loc, err = s.cache.LocateKey(s.bo, []byte("a"))
+	c.Assert(err, IsNil)
+	ctx, err := s.cache.GetTiKVRPCContext(s.bo, loc.Region, kv.ReplicaReadLeader, 0)
+	c.Assert(err, IsNil)
+	c.Assert(ctx.Peer.StoreId, Equals, s.store2)
+
+	// simulate peer1 became down in kv heartbeat and loaded before response back.
+	cpMeta := &metapb.Region{
+		Id:          ctx.Meta.Id,
+		StartKey:    ctx.Meta.StartKey,
+		EndKey:      ctx.Meta.EndKey,
+		RegionEpoch: ctx.Meta.RegionEpoch,
+		Peers:       make([]*metapb.Peer, len(ctx.Meta.Peers)),
+	}
+	copy(cpMeta.Peers, ctx.Meta.Peers)
+	cpRegion := &pd.Region{
+		Meta:      cpMeta,
+		DownPeers: []*metapb.Peer{{Id: s.peer1, StoreId: s.store1}},
+	}
+	filterUnavailablePeers(cpRegion)
+	region := &Region{meta: cpRegion.Meta}
+	err = region.init(s.cache)
+	c.Assert(err, IsNil)
+	s.cache.insertRegionToCache(region)
+
+	// OnSendFail should not panic
+	s.cache.OnSendFail(NewNoopBackoff(context.Background()), ctx, false, errors.New("send fail"))
 }
 
 func (s *testRegionRequestSuite) TestGetRegionByIDFromCache(c *C) {

